runtime/common/src/impls.rs

```diff
@@ -163,3 +166,139 @@ impl From<u128> for InstanceId {
+where
+	MaxTranches: Get<TrancheId>,
+	MaxHold: Get<Moment>,
+	MinDelay: Get<Moment>,
+{
+	type Property = PoolRole<Moment, TrancheId>;
+
+	fn exists(&self, property: Self::Property) -> bool {
+		match property {
+			PoolRole::Borrower => self.admin.contains(AdminRoles::BORROWER),
+			PoolRole::LiquidityAdmin => self.admin.contains(AdminRoles::LIQUIDITY_ADMIN),
+			PoolRole::PoolAdmin => self.admin.contains(AdminRoles::POOL_ADMIN),
+			PoolRole::PricingAdmin => self.admin.contains(AdminRoles::PRICING_ADMIN),
+			PoolRole::MemberListAdmin => self.admin.contains(AdminRoles::MEMBER_LIST_ADMIN),
+			PoolRole::RiskAdmin => self.admin.contains(AdminRoles::RISK_ADMIN),
+			PoolRole::TrancheInvestor(id, now) => self.tranche_investor.contains(id, now),
+		}
+	}
+
+	fn empty(&self) -> bool {
+		self.admin.is_empty() && self.tranche_investor.is_empty()
+	}
+
+	fn rm(&mut self, property: Self::Property) {
+		match property {
+			PoolRole::Borrower => self.admin.remove(AdminRoles::BORROWER),
+			PoolRole::LiquidityAdmin => self.admin.remove(AdminRoles::LIQUIDITY_ADMIN),
+			PoolRole::PoolAdmin => self.admin.remove(AdminRoles::POOL_ADMIN),
+			PoolRole::PricingAdmin => self.admin.remove(AdminRoles::PRICING_ADMIN),
+			PoolRole::MemberListAdmin => self.admin.remove(AdminRoles::MEMBER_LIST_ADMIN),
+			PoolRole::RiskAdmin => self.admin.remove(AdminRoles::RISK_ADMIN),
+			PoolRole::TrancheInvestor(id, now) => self.tranche_investor.remove(id, now),
+		}
+	}
+
+	fn add(&mut self, property: Self::Property) {
+		match property {
+			PoolRole::Borrower => self.admin.insert(AdminRoles::BORROWER),
+			PoolRole::LiquidityAdmin => self.admin.insert(AdminRoles::LIQUIDITY_ADMIN),
+			PoolRole::PoolAdmin => self.admin.insert(AdminRoles::POOL_ADMIN),
+			PoolRole::PricingAdmin => self.admin.insert(AdminRoles::PRICING_ADMIN),
+			PoolRole::MemberListAdmin => self.admin.insert(AdminRoles::MEMBER_LIST_ADMIN),
+			PoolRole::RiskAdmin => self.admin.insert(AdminRoles::RISK_ADMIN),
+			PoolRole::TrancheInvestor(id, now) => self.tranche_investor.insert(id, now),
+		}
+	}
+}
+
+impl<MaxHold, MinDelay, MaxTranches> TrancheInvestors<MaxHold, MinDelay, MaxTranches>
+where
+	MaxTranches: Get<TrancheId>,
+	MaxHold: Get<Moment>,
+	MinDelay: Get<Moment>,
+{
+	pub fn empty() -> Self {
+		Self::default()
+	}
+
+	pub fn is_empty(&self) -> bool {
+		self.info.is_empty()
+	}
+
+	pub fn contains(&self, tranche: TrancheId, now: Moment) -> bool {
+		if tranche >= self.max_tranches {
+			return false;
+		}
+
+		self.info
+			.iter()
+			.position(|info| {
+				info.tranche_id == tranche
+					&& info.permissioned_at.saturating_add(now) <= MaxHold::get()
+			})
+			.is_some()
+	}
+
+	pub fn remove(&mut self, tranche: TrancheId, now: Moment) {
+		if tranche >= self.max_tranches {
+			return;
+		}
+
+		if let Some(index) = self.info.iter().position(|info| {
+			info.tranche_id == tranche
+				&& info.permissioned_at.saturating_add(now) >= MinDelay::get()
```

Comment:
we should not remove the role, instead we take new `permissioned_at` that is atleast `now + MaxDelay` and update it.

We can remove the expired role through purge call if we want to save storage

Reply:
If we never remove a role, we don`t need to implement this function.

Reply:
ideally, yes, hence this will mostly be an alias for insert